### ISSUE TITLE
Type mismatch on &str

### ISSUE THREAD

**Issue by lnicola** [rust-analyzer/rust-analyzer#5020]
```rust
fn _foo(s: String) {
    match s.as_str() {
        "foo" => {}
        _ => {}
    }
}
```

This reports `Expected str, got &str`. Probably a known issue.

**Comment by flodiebold** [comment]
I think that's this:
https://github.com/rust-analyzer/rust-analyzer/blob/15c71f881607db5ddfd759785f62e4ae44aaf301/crates/ra_hir_ty/src/infer/pat.rs#L100
`"foo"` is a reference, but we don't treat it as such and apply match ergonomics. Should be easy to fix (just check whether the literal is a string literal and return false in that case).
